walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,92 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Set;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+
+  /**
+  * Queries through a list of events and creates a list of possible time ranges
+  * where the requested meeting can take place.
+  * @param events the list of events taking place to be considered.
+  * @param request a meeting request with information about the meeting.
+  * @return a list of time ranges where the requested meeting could take place.
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> possibleTimes = new ArrayList<>();
+    List<Event> eventList = new ArrayList<>(events);
+    long duration = request.getDuration();
+
+    // Begins with the assumption that the entire day is available.
+    return findOpennings(possibleTimes, eventList, 0, TimeRange.START_OF_DAY, duration, request.getAttendees());
+  }
+
+  /**
+  * Sorts the list of events taking place by their start times.
+  */
+  private static void sortEventsByStart(List<Event> eventList) {
+    eventList.sort(new Comparator<Event>() {
+      @Override
+      public int compare(Event a, Event b) {
+        return TimeRange.ORDER_BY_START.compare(a.getWhen(), b.getWhen());
+      }
+    });
+  }
+
+  /**
+  * Whether there is at least a required attendee for the meeting attending the event. 
+  */
+  private static boolean attendeesRequired(Collection<String> meetingAttendees, Set<String> eventAttendees) {
+    for (String meetingAttendee: meetingAttendees) {
+      if (eventAttendees.contains(meetingAttendee))
+        return true;
+    }
+    return false;
+  }
+
+  /**
+  * Recursively goes through the list of events to find gaps in time ranges where the
+  * requested meeting could take place.
+  */
+  private Collection<TimeRange> findOpennings(List<TimeRange> possibleTimes, List<Event> eventList, int eventIdx, int position, long duration, Collection<String> meetingAttendees) {
```

Comment:
Are you sure you need to have a recursive method? This method is just looping through all events in the event list, so it is not really a recursive algorithm.

Reply:
True. As it stands right now using recursion isn't necessary. But I might be changing it to do some recursive things. If I end up not using it I'll definitely make it iterative.